indispensable/src/commonMain/kotlin/at/asitplus/signum/indispensable/asn1/encoding/NumberEncoding.kt

```diff
@@ -241,222 +163,374 @@ fun Int.toUnsignedByteArray() = toLong().toUnsignedByteArray()
 
 /**
  * Encodes this number using varint encoding as used within ASN.1: groups of seven bits are encoded into a byte,
  * while the highest bit indicates if more bytes are to come
+ *
+ * @return the number of bytes written to the sink
  */
-fun BigInteger.toAsn1VarInt(): ByteArray {
-    if (isZero()) return byteArrayOf(0)
-    require(isPositive) { "Only positive Numbers are supported" }
-    if (this < 128) return byteArrayOf(this.byteValue(exactRequired = true)) //Fast case
+fun Sink.writeAsn1VarInt(number: UInt) = writeAsn1VarInt(number.toULong(), UInt.SIZE_BITS)
+
+/**
+ * Encodes this number using varint encoding as used within ASN.1: groups of seven bits are encoded into a byte,
+ * while the highest bit indicates if more bytes are to come
+ *
+ * @return the number of bytes written to the sink
+ */
+private fun Sink.writeAsn1VarInt(number: ULong, bits: Int):Int {
+    if (number < 128u) return writeByte(number.toByte()).run { 1 } //Fast case
     var offset = 0
     var result = mutableListOf<Byte>()
 
-    val mask = BigInteger.fromUByte(0x7Fu)
-    var b0 = ((this shr offset) and mask).byteValue(exactRequired = false)
-    while ((this shr offset > 0uL) || offset == 0) {
+    var b0 = (number shr offset and UVARINT_MASK_ULONG).toByte()
+    while ((number shr offset > 0uL) || offset == 0) {
         result += b0
         offset += 7
-        if (offset > (this.bitLength() - 1)) break //End of Fahnenstange
-        b0 = ((this shr offset) and mask).byteValue(exactRequired = false)
+        if (offset > (bits - 1)) break //End of Fahnenstange
+        b0 = (number shr offset and UVARINT_MASK_ULONG).toByte()
     }
-
-    return with(result) {
-        ByteArray(size) { fromBack(it) or asn1VarIntByteMask(it) }
+    result.forEachIndexed { index, _ ->
+        writeByte(result.fromBack(index) or result.asn1VarIntByteMask(index))
     }
+    return result.size
 }
 
 /**
- * Encodes this number using unsigned VarInt encoding as used within ASN.1:
- * Groups of seven bits are encoded into a byte, while the highest bit indicates if more bytes are to come.
+ * Encodes this number using varint encoding as used within ASN.1: groups of seven bits are encoded into a byte,
+ * while the highest bit indicates if more bytes are to come
  *
- * This kind of encoding is used to encode [ObjectIdentifier] nodes and ASN.1 Tag values > 30
+ * @return the number of bytes written to the sink
  */
-fun UInt.toAsn1VarInt(): ByteArray {
-    if (this < 128u) return byteArrayOf(this.toByte()) //Fast case
+@Throws(IllegalArgumentException::class)
+fun Sink.writeAsn1VarInt(number: BigInteger): Int {
+    if (number.isZero()) return writeByte(0).run { 1 }
+    require(number.isPositive) { "Only positive Numbers are supported" }
+    if (number < UVARINT_SINGLEBYTE_MAXVALUE) return writeByte(number.byteValue(exactRequired = true)).run { 1 } //Fast case
     var offset = 0
     var result = mutableListOf<Byte>()
```

Comment:
this function can be replaced with:
```kotlin
val numBytes = (number.bitLength()+6)/7 // division rounding up
(numBytes-1).downTo(0).forEach { byteIndex ->
    writeByte(
        ((number shr (byteIndex*7)).byteValue(exactRequired = false) and UVARINT_MASK_BIGINT) or
            (if (byteIndex > 0) UVARINT_SINGLEBYTE_MAXVALUE else 0)))
}
return numBytes
```

Reply:
already covered by the `if (number < UVARINT_SINGLEBYTE_MAXVALUE)` branch above (assuming you kept it)

Reply:
is this working now?